File: TradeItIosEmsApi/TradeItKeychain.m
//
// Keychain.h
//
// Based on code by Michael Mayo at http://overhrd.com/?p=208
//
// Created by Frank Kim on 1/3/11.
//

#import "TradeItKeychain.h"
#import <Security/Security.h>

@implementation TradeItKeychain

+ (void)saveString:(NSString *)inputString forKey:(NSString	*)account {
	NSAssert(account != nil, @"Invalid account");
	NSAssert(inputString != nil, @"Invalid string");
	
	NSMutableDictionary *query = [NSMutableDictionary dictionary];
	
	[query setObject:(__bridge id)kSecClassGenericPassword forKey:(__bridge id)kSecClass];
	[query setObject:account forKey:(__bridge id)kSecAttrAccount];
	[query setObject:(__bridge id)kSecAttrAccessibleWhenUnlocked forKey:(__bridge id)kSecAttrAccessible];
	
	OSStatus error = SecItemCopyMatching((__bridge CFDictionaryRef)query, NULL);
	if (error == errSecSuccess) {
		// do update
		NSDictionary *attributesToUpdate = [NSDictionary dictionaryWithObject:[inputString dataUsingEncoding:NSUTF8StringEncoding] 
																	  forKey:(__bridge id)kSecValueData];
		
		error = SecItemUpdate((__bridge CFDictionaryRef)query, (__bridge CFDictionaryRef)attributesToUpdate);
		NSAssert1(error == errSecSuccess, @"SecItemUpdate failed: %i", (int)error);
	} else if (error == errSecItemNotFound) {
		// do add
		[query setObject:[inputString dataUsingEncoding:NSUTF8StringEncoding] forKey:(__bridge id)kSecValueData];
		
		error = SecItemAdd((__bridge CFDictionaryRef)query, NULL);
		NSAssert1(error == errSecSuccess, @"SecItemAdd failed: %i", (int)error);
	} else {
		NSAssert1(NO, @"SecItemCopyMatching failed: %i", (int)error);
	}
}

+ (NSString *)getStringForKey:(NSString *)account {
	NSAssert(account != nil, @"Invalid account");
	
	NSMutableDictionary *query = [NSMutableDictionary dictionary];

	[query setObject:(__bridge id)kSecClassGenericPassword forKey:(__bridge id)kSecClass];
	[query setObject:account forKey:(__bridge id)kSecAttrAccount];
	[query setObject:(id)kCFBooleanTrue forKey:(__bridge id)kSecReturnData];

	NSData * dataFromKeychain = nil;

	OSStatus error = SecItemCopyMatching((__bridge CFDictionaryRef)query, (void *)&dataFromKeychain);
	
	NSString *stringToReturn = nil;
	if (error == errSecSuccess) {
        stringToReturn = [[NSString alloc] initWithData:dataFromKeychain encoding:NSUTF8StringEncoding];
	}
	
	return stringToReturn;
}

+ (void)deleteStringForKey:(NSString *)account {
	NSAssert(account != nil, @"Invalid account");

	NSMutableDictionary *query = [NSMutableDictionary dictionary];
	
	[query setObject:(__bridge id)kSecClassGenericPassword forKey:(__bridge id)kSecClass];
	[query setObject:account forKey:(__bridge id)kSecAttrAccount];
		
	OSStatus status = SecItemDelete((__bridge CFDictionaryRef)query);
	if (status != errSecSuccess) {
		NSLog(@"SecItemDelete failed: %i", (int)status);
	}
}

@end

File: TradeItIosEmsApi/TradeItKeychain.swift
class TradeItKeychain: NSObject {

    static func save(_ inputString: String?, forKey account: String?) {
        assert(account != nil, "Invalid account")
        assert(inputString != nil, "Invalid string")
        
        var query: [AnyHashable : Any] = [:]
        
        query[kSecClass] = kSecClassGenericPassword
        query[kSecAttrAccount] = account
        query[kSecAttrAccessible] = kSecAttrAccessibleWhenUnlocked
        
        var error: OSStatus = SecItemCopyMatching(query as CFDictionary, nil)
        if error == errSecSuccess {
            // do update
            var attributesToUpdate: [AnyHashable : Any] = [:]
            if let data = inputString?.data(using: .utf8) {
                attributesToUpdate = [kSecValueData : data]
            }
            
            error = SecItemUpdate(query as CFDictionary, attributesToUpdate as CFDictionary)
            assert(error == errSecSuccess, String(format: "SecItemUpdate failed: %i", Int(error)))
        } else if error == errSecItemNotFound {
            // do add
            query[kSecValueData] = inputString?.data(using: .utf8)
            
            error = SecItemAdd(query as CFDictionary, nil)
            assert(error == errSecSuccess, String(format: "SecItemAdd failed: %i", Int(error)))
        } else {
            assert(false, String(format: "SecItemCopyMatching failed: %i", Int(error)))
        }
    }
    
    static func getStringForKey(_ account: String?) -> String? {
        assert(account != nil, "Invalid account")
        
        var query: [AnyHashable : Any] = [:]
        
        query[kSecClass] = kSecClassGenericPassword
        query[kSecAttrAccount] = account
        query[kSecReturnData] = kCFBooleanTrue
        
        var dataFromKeychain: CFTypeRef? = nil
        
        let error: OSStatus = SecItemCopyMatching(query as CFDictionary, &dataFromKeychain)
        
        var stringToReturn: String? = nil
        if error == errSecSuccess {
            if let dataFromKeychain = dataFromKeychain {
                stringToReturn = String(data: dataFromKeychain.data, encoding: .utf8)
            }
        }
        
        return stringToReturn
    }
    
    static func deleteString(forKey account: String?) {
        assert(account != nil, "Invalid account")
        
        var query: [AnyHashable : Any] = [:]
        
        query[kSecClass] = kSecClassGenericPassword
        query[kSecAttrAccount] = account
        
        let status: OSStatus = SecItemDelete(query as CFDictionary)
        if status != errSecSuccess {
            print(String(format: "SecItemDelete failed: %i", Int(status)))
        }
    }
}
